ParameterTrendPercentages/ParameterTrendPercentages_1/ParameterTrendPercentages_1.cs

```diff
@@ -209,31 +208,92 @@ private void RetrieveTrendRecords()
+            Dictionary<DateTime, string> trendRecords = new Dictionary<DateTime, string>();
+
+            // Search all records before start time and add closest with updated start time
+            List<TrendRecord> beforeStartRecords = records.Where(r => r.Time <= _start).ToList();
+            trendRecords.Add(_start, beforeStartRecords.Count > 0 ? beforeStartRecords.Last().GetStringValue() : "Not trended");
 
-        // Remove empty values or outside window of request
-        List<TrendRecord> filteredRecords = records
-            .Where(r => r.Time >= _start && r.Time <= _end && !string.IsNullOrEmpty(r.GetStringValue()))
-            .ToList();
+            // Add trend records to Dictionary if value is not empty and within window
+            foreach (var record in records.Where(r => r.GetStringValue() != string.Empty && r.Time > _start && r.Time < _end))
+            {
+                // Remove milliseconds to avoid duplicate time records
+                record.Time.AddMilliseconds(record.Time.Millisecond * -1);
+
+                if (!trendRecords.ContainsKey(record.Time))
+                    trendRecords.Add(record.Time, record.GetStringValue());
+            }
 
-        // Convert the filtered records into a dictionary
-        Dictionary<DateTime, string> trendRecords = filteredRecords
-            .ToDictionary(r => r.Time, r => r.GetStringValue());
+            // Stretch last known value to end time
+            if (trendRecords.Last().Key != _end)
+                trendRecords.Add(_end, trendRecords.Last().Value);
 
-        // Count the different number of string values
-        _result = CalculateKeyDistribution(trendRecords);
+            // Calculate percentages
+            _result = Calculate(trendRecords);
+        }
+        else
+        {
+            // If no points are returned, indicate that no trending is enabled.
+            _result = new List<ResultRow>
+            {
+                new ResultRow() { Key = "Not trended", Percentage = 100, Absolute = 0 },
+            };
+        }
     }
 
-    private Dictionary<string, double> CalculateKeyDistribution(Dictionary<DateTime, string> dictionary)
+    /// <summary>
+    /// Calculate Percentages.
+    /// </summary>
+    /// <param name="records">Trend records dictionary.</param>
+    /// <returns>Dictionary with each distinct value as key and the percentage as value.</returns>
+    private List<ResultRow> Calculate(Dictionary<DateTime, string> records)
     {
-        int totalCount = dictionary.Count;
+        int totalDuration = 0;
+        var sortedRecords = records.OrderBy(entry => entry.Key).ToList();
+        Dictionary<string, int> activeTimes = new Dictionary<string, int>();
+
+        for (int i = 0; i < sortedRecords.Count - 1; i++)
+        {
+            // Calculate duration
+            int duration = (int)(sortedRecords[i + 1].Key - sortedRecords[i].Key).TotalSeconds;
```

Comment:
Could store one value outside the loop so you only need one retrieval from the array.